galaxy/api/download/views.py

```diff
@@ -0,0 +1,83 @@
+import logging
+from pathlib import Path, PurePath
+
+from django.conf import settings
+from django.core.exceptions import (
+    ImproperlyConfigured, SuspiciousFileOperation)
+from django.core.files.storage import FileSystemStorage
+from django.http import FileResponse
+from django.shortcuts import redirect, get_object_or_404
+from storages.backends import s3boto3
+from rest_framework.exceptions import NotFound
+from rest_framework.permissions import AllowAny
+
+from galaxy.api.base import APIView
+from galaxy.common.schema import CollectionFilename
+from galaxy.main import models
+
+
+__all__ = (
+    'ArtifactDownloadView',
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ArtifactDownloadView(APIView):
+    permission_classes = (AllowAny,)
+
+    def get(self, request, filename):
+        try:
+            filename = CollectionFilename.parse(filename)
+        except ValueError:
+            raise NotFound
+
+        version = get_object_or_404(
+            models.CollectionVersion,
+            collection__namespace__name__iexact=filename.namespace,
+            collection__name__iexact=filename.name,
+            version__exact=filename.version,
+        )
+
+        user_agent = request.META.get('HTTP_USER_AGENT', '')
+        if user_agent.startswith('Mazer/'):
+            version.collection.inc_download_count()
+
+        ca = version.get_content_artifact()
+        filename = PurePath(ca.relative_path).name
+
+        # If artifact url is a remote URL (e.g. S3 URL) return redirect to
+        # # this URL.
+        artifact_file = ca.artifact.file
+        storage = artifact_file.storage
+        if isinstance(storage, s3boto3.S3Boto3Storage):
+            return self._s3_redirect(artifact_file, filename)
+
+        # If running in debug mode serve file directly by Django.
+        if isinstance(storage, FileSystemStorage):
+            return self._serve_file(artifact_file, filename)
+
+        raise ImproperlyConfigured(
+            'Only S3 and local filesystem storage backends are supported.')
+
+    def _s3_redirect(self, artifact_file, filename):
+        content_disposition = f'attachment; filename={filename}'
+        parameters = {
+            'ResponseContentDisposition': content_disposition
+        }
+        url = artifact_file.storage.url(
+            artifact_file.name, parameters=parameters)
+        return redirect(url)
+
+    def _serve_file(self, artifact_file, filename):
```

Comment:
These could be defined outside the class or as static methods, but I'm not picky about that so I would leave them as is (especially since nothing external to the class uses them).

Reply:
These functions are tightly related to the view class and not intended to be reused elsewhere, therefore these better to be kept in a view class scope.